common/src/main/java/com/wynntils/wrappedscreens/trademarket/TradeMarketSearchResultHolder.java

```diff
@@ -0,0 +1,370 @@
+        ContainerUtils.clickOnSlot(
+                BACK_TO_SEARCH_SLOT,
+                wrappedScreenInfo.containerId(),
+                GLFW.GLFW_MOUSE_BUTTON_LEFT,
+                wrappedScreenInfo.containerMenu().getItems());
+    }
+
+    public void loadNextPageBatch() {
+        if (allPagesLoaded) return;
+
+        runOrQueueAction(new QueuedAction(PageLoadingMode.LOAD_ITEMS, itemMap.size() - 1 + PAGE_BATCH_SIZE, null));
+    }
+
+    public void updateDisplayItems(ItemSearchQuery searchQuery) {
+        filteredItems.clear();
+
+        Stream<ItemStack> items = itemMap.values().stream()
+                .map(map -> map.values().toArray(new ItemStack[0]))
+                .flatMap(Stream::of);
+
+        List<ItemStack> matchingItems = Services.ItemFilter.filterAndSort(searchQuery, items.toList());
+
+        filteredItems.addAll(matchingItems);
+    }
+
+    public List<ItemStack> getFilteredItems() {
+        return filteredItems;
+    }
+
+    public int getPageLoadBatchSize() {
+        return PAGE_BATCH_SIZE;
+    }
+
+    private void handleSetItem(int slot, ItemStack itemStack) {
+        // We only care about the slots that are in the "search results" area
+        if (slot % 9 >= 7 || slot >= LAST_ITEM_SLOT) return;
+
+        // If we have the item count we expect on the page, we can load the next page
+        // If we have found an empty item, we count them and check if we have the expected amount
+        Int2ObjectOpenHashMap<ItemStack> currentItems =
+                itemMap.computeIfAbsent(this.currentPage, k -> new Int2ObjectOpenHashMap<>());
+
+        // Remove the item if it was already there,
+        // if it is not an empty item, we add it back
+        currentItems.remove(slot);
+
+        boolean emptyItem = isEmptyItem(itemStack);
+        if (!emptyItem) {
+            // Update item in slot, when changing pages,
+            // items can change
+            ItemStack oldItem = currentItems.put(slot, itemStack);
+        }
+
+        pageItemCount++;
+
+        switch (pageLoadingMode) {
+            case LOAD_ITEMS -> pageLoadedWhileLoadingItems(currentItems);
+            case CLICK_ITEM -> pageLoadedWhileSelectingItem(currentItems, slot, itemStack);
+        }
+    }
+
+    private void pageLoadedWhileLoadingItems(Int2ObjectOpenHashMap<ItemStack> currentItems) {
+        wrappedScreen.setCurrentState(Component.literal("Loading page " + (currentPage + 1) + "..."));
+
+        // We only go to the next page if we have the expected amount of items
+        if (pageItemCount != EXPECTED_ITEMS_PER_PAGE) return;
+
+        // If we have air items on the page, we reached the end
+        if (currentItems.size() < EXPECTED_ITEMS_PER_PAGE) {
+            wrappedScreen.setCurrentState(Component.literal("All pages loaded"));
+
+            allPagesLoaded = true;
+
+            pageLoadingMode = PageLoadingMode.NONE;
+            requestedPage = -1;
+            runQueuedAction();
```

Comment:
These lines seem to be duplicated several times. Perhaps extract to a separate method?